service/history/hsm/tree.go

```diff
@@ -327,6 +337,45 @@ func (n *Node) AddChild(key Key, data any) (*Node, error) {
 	return node, nil
 }
 
+// DeleteChild deletes an immediate child node and all its descendants from the tree.
+func (n *Node) DeleteChild(key Key) error {
+	if n.cache.deleted {
+		return fmt.Errorf("%w: cannot delete from deleted node: %v", ErrStateMachineInvalidState, n.Key)
+	}
+
+	child, err := n.Child([]Key{key})
+	if err != nil {
+		return fmt.Errorf("failed to get child: %w", err)
+	}
+
+	// Mark entire subtree as deleted
+	if err := child.Walk(func(n *Node) error {
+		n.cache.deleted = true
+		return nil
+	}); err != nil {
+		return err
+	}
+
+	// Record deletion
+	n.cache.outputs = append(n.cache.outputs, TransitionOutputWithCount{
```

Comment:
I would keep the op log instead of recording in the parent node.
Note that you're marking the parent as deleted here instead of the child.

Reply:
I'm trying to direct you to add transitions to the op log, and use the op log to power the `Output` function. You did the opposite.